### PR TITLE
Add autocontrast_preserve

### PR DESCRIPTION
**Pull request by broxeph**
Maximize (normalize) image contrast while preserving image tone. (Mimics
Photoshop autocontrast functionality as opposed to the default PIL
implementation, which operates on channels separately).

### CODE REVIEW

**Comment by hugovk** [comment]
Please could you add tests for this?

```
PIL/ImageOps.py (1.7%): Missing lines 150-156,158-160,162-164,167-169,171-175,177-180,182-186,188-191,193-200,202-204,206,208-218
-------------
Total:   58 lines
Missing: 57 lines
Coverage: 1%
```

https://travis-ci.org/python-pillow/Pillow/jobs/64760914#L3778

Docs would be very welcome too.

**Comment by broxeph** [comment]
Not quite sure what you're looking for in terms of docs, but I added "while preserving image tone" to a copy of the autocontrast docstring.

**Comment by wiredfool** [comment]
The tests that were added do show that it doesn't crash, but don't validate correctness.

Possible Additional tests:
- GS ramp of 256 greys, cutoff n% do we get the expected number of full white/black pixels.
- Single color image, grey, black, white. (or single channel.)
- Synthetic image with ramp in one channel +constant value + light noise in the other channels.

File: PIL/ImageOps.py [hunk 1]
```diff
@@ -135,6 +135,89 @@ def autocontrast(image, cutoff=0, ignore=None):
...
+                h[ignore] = 0
+            except TypeError:
+                # assume sequence
+                for ix in ignore:
+                    h[ix] = 0
+        if cutoff:
+            # cut off pixels from both ends of the histogram
+            # get number of pixels
+            n = 0
+            for ix in range(256):
```
Comment (wiredfool) [inline PIL/ImageOps.py]: n = sum(h)?

File: PIL/ImageOps.py [hunk 2]
```diff
@@ -135,6 +135,89 @@ def autocontrast(image, cutoff=0, ignore=None):
...
+    # get lowest/highest samples across all layers
+    lo = min(los)
+    hi = max(his)
+    if hi <= lo:
+        # don't bother
+        lut.extend(list(range(len(histogram))))
+    else:
+        scale = 255.0 / (hi - lo)
+        offset = -lo * scale
+        for layer in range(len(histogram) // 256):
```
Comment (wiredfool) [inline PIL/ImageOps.py]: This should create either [lut] or [lut, lut, lut], calculating it 3x isn't necessary. Even appending it 3x isnt' necessary, due to the implementation of `_lut`.

**Comment by wiredfool** [comment]
I've added some tests here (https://github.com/wiredfool/Pillow/commit/bb3ad2750a8055bbffea7812ffa0e2f43ad79a18) for correctness of implementation, and I'm seeing some really strange behavior for running against a constant color image. 

If all pixels are white, it returns a white image. (127, 255,0) returns the same.  
But (127,127,127) returns an image that's (127, 255, 255), and an all black image returns (0,255,255).

**Comment by hugovk** [comment]
I'll close this because there's been no progress in 3-4 years. Ping if you'd like to re-open it, or create a new PR. Thanks!

**Comment by radarhere** [comment]
An alternative to this, #5350, has been merged

**Pull request closed without merging**